### ISSUE TITLE
Convert x & y coordinates into real world measurements?

### ISSUE THREAD

**Issue by syu-id** [isolver/OpenHandWrite#226]
Hi. We want to calculate the length of a stroke by using the coordinates of the sample points.

For example, we measure the length of the vertical stroke highlighted in the picture below. By adding up the distance between every two consecutive samples, we get a length of 2574. But how do we translate this into a real world measurement such as millimetres?

Can one convert the coordinates into real world measurements based on the size & resolution of a tablet PC's screen or a pen tablet's writing area? I am currently using a 2-in-1 PC (14inch ThinkPad X1 Yoga).

![strokelength](https://user-images.githubusercontent.com/1129506/108700257-aff40780-7549-11eb-91fe-3ff405ffea6f.png)

**Comment by Mark-Torrance** [comment]
I'm assigning this to Guido @g-no who will be able to give you a definitive answer. But this should be a simple conversion, using the lines-per-cm resolution for the Yoga digitiser. For the Wacom tablet that we've been using this is 1300 (from memory).

[Tangentially, it's really good to know that this works on Thinkpad. I didn't realise that they used Wintab.]

**Comment by g-no** [comment]
I'll look into this and post an answer later today.

**Comment by syu-id** [comment]
Thank you Mark and Guido!
[And yes, OpenHandWrite works on Thinkpad Yoga nicely!]

**Comment by g-no** [comment]
You need to find out the spatial resolution of your digitizer (this is not the same as your LCD resolution). For example an Intuos4 XL A3 Wide has 5,080 lines per inch, i.e. 1000 lines per cm == 0.001 cm/line. You can find out the resolution of your device by starting a new session and move the pen to the ends of all sides (top/bottom/left/right). In the data you look for the x min/max (min should be 0) and y min/max (again min should be zero). Next you measure the physical height and width and then you can convert your data to millimeter. 
If you max(x) is e.g. 15450 and your screen is 309 mm wide then your x resolution is 50 lpmm (lines per mm). Does this help?
If you collect the data using GetWrite you can specify these details in the preferences and get the mm directly in MarkWrite.

**Comment by Mark-Torrance** [comment]
Let us know what you find. Quick googling hasn't come up with any official specs.

**Comment by syu-id** [comment]
Thank you so much for your suggestions!
So for pen tablets, we just need to check out its specification sheet and find out its lines per inch.

For my PC, I looked up the physical size for a 16:9 14inch display here: http://screen-size.info/.
The dimensions should typically be 310mm x 174mm.

The max x and y values I collected were roughly 31000 and 17400.
So my 2017 Thinkpad Yoga should have a lpmm=100.
Then the length of the vertical stroke shown in my first picture should be 2574/100 = 25.74mm.

<img src="https://user-images.githubusercontent.com/1129506/108820179-2a329380-75ff-11eb-88cf-5db8313b1047.png" width="600">

**Comment by Mark-Torrance** [comment]
If you are using Intuos tablets I think they are 130 lpmm. The official spec is 100, but we observed130 and I think @s-he confirmed this.

**Comment by syu-id** [comment]
Thank you! So the official spec may not always be accurate ...
We will collect the max x and y values and calculate lpi or lpmm for pen tablets too.

**Comment by Mark-Torrance** [comment]
In this case they actually changed the spec but for some reason didn't advertise the fact. Stefan can confirm this though (I might be misremembering).

**Comment by syu-id** [comment]
I see ... So it's always the best to confirm the actual lpmm by oneself to avoid such problems.
Thank you!

**Comment by s-he** [comment]
According to the manual, the spatial resolution of Intuos 4 tablets is 200lpmm. As far as an Intuos 4XL tablet is concerned, we, however, observed a spatial resolution of 100lpmm.

To verify your calculations, you could just produce some horizontal, vertical, or diagonal strokes of e.g., 1 or 10 cm (by means of a ruler).

**Comment by syu-id** [comment]
Thank you Stefan!
Draw lines with a ruler is a nice way to verify the calculation!

**Comment by Mark-Torrance** [comment]
I clearly just made 1300 up then! For plug-in tablet like Intuos there is also the issue of what the resolution is set to in the Wacom software, of course. Measure it for yourself is clearly the way to go :)

**Comment by s-he** [comment]
“For plug-in tablet like Intuos there is also the issue of what the resolution is set to in the Wacom software, of course.”

This holds for the temporal resolution, too. You could set it to about 200Hz, but also to a lower temporal resolution (as fas as Wacom Intuos tablets are concerned). I cannot specify the lower temporal resolution at the moment, but I guess it is around 133 Hz. Typically, build-in tablets have a lower temporal resolution than plug-in tablets.

So, you also need to check your temporal resolution to correctly calculate variables such as duration, velocity etc.

**Comment by Mark-Torrance** [comment]
[Edits to remove email chain, because this is a public repository.]

GetWrite accesses sample times directly (i.e. from the computer clock). So I don't think duration and velocity are dependent on temporal resolution.

**Comment by syu-id** [comment]
It's great to know that GetWrite accesses the computer clock directly.
Thank you very much for all your kind help and information, I think my problem is successfully solved!

**Comment by Mark-Torrance** [comment]
The reason MarkWrite needs to know about temporal resolution is because this is necessary for determining when the pen is out of range. You need to know when the next sample should come in order to decide whether or not the pen is still sampling. This is also important when MarkWrite analyses data that have been captured using programs other than GetWrite.

**Comment by syu-id** [comment]
So we don't need to worry about the accuracy of timestamps if the samples are collected with GetWrite. But it is important to know the temporal resolution if the samples come from other programs and computer clock time is not recorded. Is my understanding correct? Thank you!

**Comment by Mark-Torrance** [comment]
> So we don't need to worry about the accuracy of timestamps if the samples are collected with GetWrite. 

True. Timestamp accuracy will be best-possible within the PsychoPy / ioHUB environment in which GetWrite is built. So "industry standard" accuracy.

> But it is important to know the temporal resolution if the samples come from other programs and computer clock time is not recorded. Is my understanding correct? Thank you!

No. Data from other programs will have timestamps as well. MarkWrite needs time, x, y and pressure (which can be 1,0) to work. It only needs to know sampling rate (temporal resolution) so that if knows when a sample has been missed, and therefore knows when the pen is out of range. This is needed so that when you are marking up you can, for example, skip to the next point where the pen was lifted (in data where zero pressure samples were not recorded) or lifted out of recording range (in GetWrite data). I don't know how MarkWrite uses user-provided sampling rate information. But if you don't provide this, then it uses modal inter-sample interval. See here: https://github.com/isolver/OpenHandWrite/wiki/MarkWrite-Series-detection

**Comment by syu-id** [comment]
Thank you so much for your detailed answers! Now I understand that sampling rate is crucial for detecting "series boundaries". It's also great to know that we don't need to worry about timestamp accuracy when using GetWrite within the Psychopy / ioHub environment.

**Comment by isolver** [comment]
I'm pretty sure we can read the max raw output range a device will use from the device itself. I'd have to double check that my memory is correct though. If it is, we could add params to the wintab device config that allow a user to specify the width and height of the digitizer surface. If provided, we could convert raw units into the width / height units specified in the config. 

@Mark-Torrance , @g-no, is this something we should look into adding, or atleast add to the issue list as an idea?

**Comment by isolver** [comment]
You can read the wintab device's native resolution / reporting range, using the `getHardwareConfig()` method of the iohub wintab device. Once iohub has started you can access the wintab device:

```
# io is the iohub class in your experiment
tablet = io.devices.tablet
```

and then read the hardware config iohub collected from the wintab deivce driver, including the x and y output range (as reported by the wintab hardware, so should be verified):

```
wintab_axis_info = tablet.getHardwareConfig().get('Axis')
x_axis = wintab_axis_info.get('x')
y_axis = wintab_axis_info.get('y')
print "X:", x_axis.get('min'), x_axis.get('max')
print "Y:", y_axis.get('min'), y_axis.get('max')
```

outputs the following for the PTK-440:

```
X: 0 31495
Y: 0 19684
```

I wonder if this could help to automate conversions from the native wintab resolution and the experiment display resolution.
